app/post.go

```diff
@@ -448,6 +448,90 @@ func (a *App) UpdatePost(post *model.Post, safeUpdate bool) (*model.Post, *model
 	}
 }
 
+func (a *App) RethreadPost(post *model.Post, safeUpdate bool) (*model.Post, *model.AppError) {
+	post.SanitizeProps()
+
+	var oldPost *model.Post
+	var thread *model.PostList
+	if result := <-a.Srv.Store.Post().Get(post.Id); result.Err != nil {
+		return nil, result.Err
+	} else {
+		oldPost = result.Data.(*model.PostList).Posts[post.Id]
+		thread = result.Data.(*model.PostList)
+		if oldPost == nil {
+			err := model.NewAppError("UpdatePost", "api.post.rethread_post.find.app_error", nil, "id="+post.Id, http.StatusBadRequest)
+			return nil, err
+		}
+
+		if oldPost.DeleteAt != 0 {
+			err := model.NewAppError("UpdatePost", "api.post.rethread_post.permissions_details.app_error", map[string]interface{}{"PostId": post.Id}, "", http.StatusBadRequest)
+			return nil, err
+		}
+
+		if oldPost.IsSystemMessage() {
+			err := model.NewAppError("UpdatePost", "api.post.rethread_post.system_message.app_error", nil, "id="+post.Id, http.StatusBadRequest)
+			return nil, err
+		}
+	}
+
+	var originalRootId string
+	originalRootId = oldPost.RootId
+	newPost := &model.Post{}
+	*newPost = *oldPost
+
+	if post.RootId != "" && post.RootId != newPost.Id {
```

Comment:
If I'm interpreting these correctly, we only allow changing the root to another post if this post has no replies.

What about the following cases:
* unrooting a reply from an existing thread, i.e. setting `newPost.RootId = post.RootId`
* merging two threads, i.e. `len(thread.Posts) > 1`
* preventing thread merging if the posts are in separate channels (or allowing that, effectively allowing a post to be moved to another channel as well)
* preventing thread nesting: i.e. `getPost(post.RootId) != post.RootId`

Reply:
Rethreading is also allowed if the post is already in a thread but isn't the root post. The root post can be rethreaded only when all its replies are rethreaded

Merging and changing channels are not currently implemented. Rethreading to another channels post is not possible in the UI, though now that I think about it I suppose it is possible to do the rethread to any post anyway if the user knows how to send a correct request. It should be easy to add a check that the post and root are on the same channel. 
Merging threads would require making edits to all the posts in the thread